docs/connections.md

```diff
@@ -0,0 +1,597 @@
+It is allowed only for method overloads that differ in the type of the `this`
+parameter. This limitation may be relaxed in the future, if proper motivating use-cases are found.
+
+The identifier is always accessible scope-wide, both before and after the line
+introducing it. Some monadic contexts (like IO) can introduce order-dependent
+behavior, however. This is not something that IDE is (or can be) concerned about
+when figuring out connections.
+
+Identifier is bound by using a variable name in the pattern context. Exact behavior depends on the language construction that was to introduce the identifier.
+
+Identifier introduced into a scope is visible only in the scope's subtree
+(lexical scoping). 
+
+Scopes, in the core language, are introduced by:
+* module (file), i.e. the root scope;
+* code blocks, i.e. indented blocks that follow a trailing operator;
+* arrow (lambda) `->` operator for its right-hand side.
+
+Some other constructs seemingly introduce scope (like function definitions) but
+this is because they are desugared into some construct that introduces scope (like lambdas).
+
+TODO: Consider if there are any special rules for signatures on definitions, or
+is this just type ascription lying next to a definition.
+
+### Examples
+Example:
+```
+main =
+    succ = a -> a+1
+    succ 0
+```
+
+The sample code is a module with a single top-level definition. 
+
+Here we have four scopes, each one nested in the previous one:
+* root scope (consisting of unindented lines), where `main` is visible;
+* definition and code block (all lines equally indented after
+  `main =`) scopes, where both `main` and `succ` are visible
+* lambda body scope (right-hand side after `a ->`), where all `main`,
+  `succ` and `a` are visible.
+
+Example:
+```
+test = a -> b ->
+    sum = a + b
+```
+
+Here we have root scope, then scopes of lambdas (after `a ->`
+and after `b ->`) and finally scope for code block.
+
+
+Example:
+
+```
+main = 
+    foo : a
+    foo = 2
+
+    bar : a
+    bar = 3
+```
+
+While `main` as root-level has its own scope (as a definition in root it is
+treated as method and desugared to lambda), `foo` and `bar` do not. `a`
+introduced by type signatures belongs to the `main`'s scope, and is shared by
+both nested definitions.
+
+## Patterns
+Pattern contexts are spans in the code where variables can be used to introduce
+new identifiers into some scope. Constructors names are used to pattern match
+against and potentially destructure more complex values.
+
+The following spans are pattern contexts:
+* left-hand side of assignment operator, e.g. `main` in  `main = println
+  "Hello"`;
+* right-hand side of a colon operator, e.g. `a` in `foo:a`;
+* left-hand side of an arrow operator, e.g. `a` in `a -> a + 1`.
+
+Details will follow with description of these operators.
+
+// TODO What about `case … of` ? Are there any other pattern-introducing constructs?
```

Comment:
The scrutinee of a case expression is _not_ a pattern context, but the left hand side of the battern branches is.